### DISCUSSION TITLE
Is it possible to track atom index back to SMILES string index？

### DISCUSSION THREAD

**Discussion by HelloJocelynLu** [rdkit/rdkit#4151] [Q&A] [answered]
Hi there,
Now I'm working on a text-based model and get some text-based properties in terms of SMILES. I want to visualize/analyze the results based on molecular graph. That's why I want to find a map from atom index to string index.
**Example**:
Suppose I have a molecule: `C[C@@H](Cl)C(=O)c1ccccc1`
Now I get a Mol object with 11 atoms, what I want to get is a map: atom_idx -> str_idx, i.e.:
def map_atom2str("`C[C@@H](Cl)C(=O)c1ccccc1`") -> dict:
{0: (0, 1), 1: (1, 7), 2: (8, 10), 3: (11, 12), 4: (14, 15), 5: (16, 17), 6: (18, 19), 7: (19, 20), 8: (20, 21), 9: (21, 22), 10: (22, 23)}
That means for the atom 0, it is corresponding to smiles[0:1] => `C`; for the atom 1, it is corresponding to smiles[1:7] => `[C@@H]`...etc
Currently I'm trying:
```python
def map_atom2str(smiles):
    mol = Chem.MolFromSmiles(smiles)
    atom_symbols = [s.GetSymbol().lower() for s in mol.GetAtoms()]
    map_dict = {}
    j = 0
    for i,sym in enumerate(atom_symbols):
        while smiles[j:j+len(sym)].lower() != sym:
            j += 1
        map_dict[i] = (j, j+len(sym))
        j += len(sym)
    return map_dict
```
But obviously it simply matches element symbols. Therefore, SMILES unit such as `[C@@H]`, `[Na+]`,...etc cannot be identified. In fact, I'm also very interested in bond index tracking, such as {3:(13,14)} whenever possible. Is it possible to do it in python?
Any helps would be appreciated!

**Comment by greglandrum** [comment, marked as answer]
The RDKit doesn't provide any way to access this information.

Here's a gist which attempts to solve the problem of locating the atoms with a regex:
https://gist.github.com/greglandrum/ad6ec0e9bc3272cd891319f1d81b2686
Given that a regex is involved, there's probably something this is missing, but it ought to be enough to help you get started.

Identifying the bonds, which are often not actually present in the SMILES, is considerably more difficult. The RDKit tracks the SMILES order of the bonds internally when parsing the SMILES, but this information is then discarded after the molecule is constructed.

Another thing to look at: Andrew Dalke's smiview may have the pieces you need to build something yourself:
https://hg.sr.ht/~dalke/smiview/browse/README.rst?rev=tip

**Reply by HelloJocelynLu** [reply]
A simple regex may not be able to find correct atoms if I have a molecule like "Cc1ccccc1" (end up with `Cc`). But you are super helpful and now I have some ideas to do that! Many thanks!!

**Reply by AlanKerstjens** [reply]
When working with canonical SMILES, is it guaranteed that SMILES characters will appear in the same order as the atom indices?

**Reply by greglandrum** [reply]
@AlanKerstjens assuming you are talking about parsing the SMILES: the ordering of the atoms in the molecule is the same as the ordering of the atoms in the input SMILES. If there are no explicit H atoms (i.e. `[H]` tokens) *or* if the removeHs option to the SMILES parser is set to `False`, then the indices will also be the same.

**Reply by AlanKerstjens** [reply]
@greglandrum I phrased the question poorly. I'm interested in the reverse scenario: if I have a RDKit molecule object and I write out its canonical SMILES, will the SMILES characters follow the order of the atom indices?

**Reply by ptosco** [reply]
@AlanKerstjens When you write out canonical SMILES, a `_smilesAtomOutputOrder` property is added to your RDKit molecule with the SMILES->mol index correspondences, which will be 1:1 if the atom order was already canonical:
https://github.com/rdkit/rdkit/issues/794#issuecomment-189579302

**Reply by AlanKerstjens** [reply]
Thanks @ptosco! `_smilesAtomOutputOrder` is what I was looking for. According to your last statement, if I renumber the atoms with the canonical ranks, I'd expect the non-canonical and canonical SMILES to be the same. However, this doesn't seem to be the case. This was also noted in this issue: https://github.com/rdkit/rdkit/issues/2637. Am I mixing up things?

```cpp
#include <GraphMol/SmilesParse/SmilesParse.h>
#include <GraphMol/SmilesParse/SmilesWrite.h>
#include <GraphMol/new_canon.h>
#include <GraphMol/MolOps.h>

int main() {
  RDKit::ROMol* molecule = RDKit::SmilesToMol("BrCCC(CCO)CCN");

  std::vector<unsigned> canonical_atom_order;
  RDKit::Canon::rankMolAtoms(*molecule, canonical_atom_order);
  RDKit::ROMol* canonical_molecule = RDKit::MolOps::renumberAtoms(*molecule, canonical_atom_order);

  std::cout << "Canonical SMILES: " << RDKit::MolToSmiles(*molecule) << std::endl;
  std::cout << "Non-canonical SMILES of canonical molecule: " << RDKit::MolToSmiles(*canonical_molecule, true, false, -1, false) << std::endl;

  return 0;
};
```

**Reply by ptosco** [reply]
@AlanKerstjens That's because `RDKit::Canon::rankMolAtoms` yields the order that each atom in the current molecule will have in the canonical structure, while `RDKit::MolOps::renumberAtoms` expects to operate on `_smilesAtomOutputOrder`, i.e. the original index of the atom in the molecule after canonicalization.
So you only need to invert the mapping as in the Python code below:
```
from rdkit import Chem
mol = Chem.MolFromSmiles("BrCCC(CCO)CCN")
canonical_atom_order = Chem.CanonicalRankAtoms(mol)
tuple(canonical_atom_order)
(2, 5, 8, 9, 7, 4, 1, 6, 3, 0)
canonical_atom_order_inverted = tuple(zip(*sorted((j, i) for i, j in enumerate(canonical_atom_order))))[1]
canonical_atom_order_inverted
(9, 6, 0, 8, 5, 1, 7, 4, 2, 3)
canonical_mol = Chem.RenumberAtoms(mol, canonical_atom_order_inverted)
Chem.MolToSmiles(mol, canonical=True), Chem.MolToSmiles(canonical_mol, canonical=False)
('NCCC(CCO)CCBr', 'NCCC(CCO)CCBr')
```

**Reply by AlanKerstjens** [reply]
@ptosco That makes sense. Thanks a lot for the help!
